src/Core/System/Consent/Service/ConsentService.php

```diff
@@ -0,0 +1,211 @@
+<?php declare(strict_types=1);
+
+namespace Shopware\Core\System\Consent\Service;
+
+use Psr\EventDispatcher\EventDispatcherInterface;
+use Shopware\Core\Framework\Log\Package;
+use Shopware\Core\System\Consent\ConsentContext;
+use Shopware\Core\System\Consent\ConsentDefinition;
+use Shopware\Core\System\Consent\ConsentException;
+use Shopware\Core\System\Consent\ConsentRepository;
+use Shopware\Core\System\Consent\ConsentScope;
+use Shopware\Core\System\Consent\ConsentStatus;
+use Shopware\Core\System\Consent\DTO\ConsentState;
+use Shopware\Core\System\Consent\DTO\ConsentStateLogRecord;
+use Shopware\Core\System\Consent\Event\ConsentAcceptedEvent;
+use Shopware\Core\System\Consent\Event\ConsentRevokedEvent;
+
+/**
+ * @internal
+ */
+#[Package('data-services')]
+class ConsentService
+{
+    /**
+     * @var array<string, ConsentDefinition>
+     */
+    private array $consentDefinitions;
+
+    /**
+     * @var array<string, ConsentState>
+     */
+    private ?array $states = null;
+
+    /**
+     * @param iterable<ConsentDefinition> $consentDefinitions
+     */
+    public function __construct(
+        iterable $consentDefinitions,
+        private readonly ConsentRepository $consentRepository,
+        private readonly EventDispatcherInterface $eventDispatcher
+    ) {
+        $definitions = [];
+        foreach ($consentDefinitions as $definition) {
+            $definitions[$definition->getName()] = $definition;
+        }
+        $this->consentDefinitions = $definitions;
+    }
+
+    /**
+     * @return array<ConsentState>
+     */
+    public function list(ConsentContext $context): array
+    {
+        $states = $this->fetchStates();
+
+        return array_map(function (ConsentDefinition $consent) use ($context, $states) {
+            $identifier = $context->getIdentifierForScope($consent->getScope());
+            $key = $this->key(
+                $consent->getName(),
+                $consent->getScope()->value,
+                $identifier
+            );
+
+            return $states[$key] ?? new ConsentState(
+                name: $consent->getName(),
+                scope: $consent->getScope(),
+                identifier: $identifier,
+                status: ConsentStatus::REQUESTED,
+                actorId: null,
+            );
+        }, $this->consentDefinitions);
+    }
+
+    public function getConsentState(string $name, ?string $identifier = null): ConsentState
```

Comment:
similiar here, the caller needs to know only the name of the consent definition, but also if a identifier is required and what the identifier is, that is bad design IMHO, instead you should only pass the name and the context, here we get the definition by the name and pass the context to the definition to figure out the correct scope identifier for that consent